File: WebDriver-Series/WebDriver.Series.Tests/GridTestCases/OrderIdColumnAsserter.cs
// <copyright file="OrderIdColumnAsserter.cs" company="Automate The Planet Ltd.">
// Copyright 2016 Automate The Planet Ltd.
// Licensed under the Apache License, Version 2.0 (the "License");
// You may not use this file except in compliance with the License.
// You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
// <author>Anton Angelov</author>
// <site>http://automatetheplanet.com/</site>
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using WebDriver.Series.Tests.Controls;
using WebDriver.Series.Tests.GridTestCases.Pages;

namespace WebDriver.Series.Tests.GridTestCases
{
    public class OrderIdColumnAsserter : GridColumnAsserter
    {
        public OrderIdColumnAsserter(IGridPage gridPage) : base(gridPage)
        {
        }

        public void OrderIdEqualToFilter()
        {
            this.GridPage.NavigateTo();
            var newItem = this.CreateNewItemInDb();

            this.GridPage.Grid.Filter(GridColumns.OrderID, Enums.FilterOperator.EqualTo, newItem.OrderId.ToString());           
            this.WaitForGridToLoad(1, this.GridPage.Grid);
            var items = this.GridPage.Grid.GetItems<GridItem>();
        
            Assert.AreEqual(1, items.Count);
        }
        
        public void OrderIdGreaterThanOrEqualToFilter()
        {
            this.GridPage.NavigateTo();

            // Create new item with unique ship name;
            var newItem = this.CreateNewItemInDb();
            // Create second new item with the same unique shipping name 
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            
            // When we filter by the second unique column ShippingName, only one item will be displayed. Once we apply the second not equal to filter the grid should be empty.
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.GreaterThanOrEqualTo, newItem.OrderId.ToString()),
                new GridFilter(GridColumns.ShipName, Enums.FilterOperator.EqualTo, newItem.ShipName));
            this.WaitForGridToLoadAtLeast(2, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();

            Assert.AreEqual(secondNewItem.OrderId, results.FirstOrDefault(x => x.ShipName == secondNewItem.ShipName).OrderId);
            Assert.AreEqual(newItem.OrderId, results.FirstOrDefault(x => x.ShipName == newItem.ShipName).OrderId);
            Assert.IsTrue(results.Count() == 2);
        }
        
        public void OrderIdGreaterThanFilter()
        {
            this.GridPage.NavigateTo();
            
            // Create new item with unique ship name;
            var newItem = this.CreateNewItemInDb();
            // Create second new item with the same unique shipping name 
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            
            // Filter by the smaller orderId but also by the second unique column in this case shipping name
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.GreaterThan, newItem.OrderId.ToString()),
                new GridFilter(GridColumns.ShipName, Enums.FilterOperator.EqualTo, newItem.ShipName));
            this.WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();
            
            Assert.AreEqual(secondNewItem.OrderId, results.FirstOrDefault(x => x.ShipName == secondNewItem.ShipName).OrderId);
            Assert.IsTrue(results.Count() == 1);
        }
        
        public void OrderIdLessThanOrEqualToFilter()
        {
            this.GridPage.NavigateTo();
            
            // Create new item with unique ship name;
            var newItem = this.CreateNewItemInDb();
            // Create second new item with the same unique shipping name 
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            
            // Filter by the larger orderId but also by the second unique column in this case shipping name
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.LessThanOrEqualTo, secondNewItem.OrderId.ToString()),
                new GridFilter(GridColumns.ShipName, Enums.FilterOperator.EqualTo, newItem.ShipName));
            this.WaitForGridToLoadAtLeast(2, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();
            
            Assert.AreEqual(newItem.OrderId, results.FirstOrDefault(x => x.ShipName == newItem.ShipName).OrderId);
            Assert.AreEqual(secondNewItem.OrderId, results.Last(x => x.ShipName == secondNewItem.ShipName).OrderId);
            Assert.IsTrue(results.Count() == 2);
        }
        
        public void OrderIdLessThanFilter()
        {
            this.GridPage.NavigateTo();
            
            // Create new item with unique ship name;
            var newItem = this.CreateNewItemInDb();
            // Create second new item with the same unique shipping name 
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            
            // Filter by the larger orderId but also by the second unique column in this case shipping name
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.LessThan, secondNewItem.OrderId.ToString()),
                new GridFilter(GridColumns.ShipName, Enums.FilterOperator.EqualTo, secondNewItem.ShipName));
            this.WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();
            
            Assert.AreEqual(newItem.OrderId, results.FirstOrDefault(x => x.ShipName == newItem.ShipName).OrderId);
            Assert.IsTrue(results.Count() == 1);
        }
        
        public void OrderIdNotEqualToFilter()
        {
            this.GridPage.NavigateTo();
            
            // Create new item with unique ship name;
            var newItem = this.CreateNewItemInDb();
            // Create second new item with the same unique shipping name 
            var secondNewItem = this.CreateNewItemInDb(newItem.ShipName);
            
            // Filter by the larger orderId but also by the second unique column in this case shipping name
            this.GridPage.Grid.Filter(
                new GridFilter(GridColumns.OrderID, Enums.FilterOperator.NotEqualTo, secondNewItem.OrderId.ToString()),
                new GridFilter(GridColumns.ShipName, Enums.FilterOperator.EqualTo, secondNewItem.ShipName));
            this.WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();
            
            Assert.AreEqual(newItem.OrderId, results.FirstOrDefault(x => x.ShipName == newItem.ShipName).OrderId);
            Assert.IsTrue(results.Count() == 1);
        }
        
        public void OrderIdClearFilter()
        {
            this.GridPage.NavigateTo();
            // Create new item with unique ship name;
            var newItem = this.CreateNewItemInDb();
            // Make sure that we have at least 2 items if the grid is empty. The tests are designed to run against empty DB.
            this.CreateNewItemInDb(newItem.ShipName);
            
            this.GridPage.Grid.Filter(GridColumns.OrderID, Enums.FilterOperator.EqualTo, newItem.OrderId.ToString());
            this.WaitForGridToLoad(1, this.GridPage.Grid);
            this.GridPage.Grid.RemoveFilters();
            
            this.WaitForGridToLoadAtLeast(1, this.GridPage.Grid);
            var results = this.GridPage.Grid.GetItems<Order>();
            Assert.IsTrue(results.Count() > 1);
        }
    }
}

File: java/WebAutomation-Series/src/test/java/kendogrid/reuse/OrderIdColumnAsserter.java
package kendogrid.reuse;

import kendogrid.Order;
import kendogrid.components.FilterOperator;
import kendogrid.components.GridFilter;
import kendogrid.components.GridItem;
import org.openqa.selenium.WebDriver;
import org.testng.Assert;
import org.testng.annotations.Test;

import java.util.List;

public class OrderIdColumnAsserter extends GridColumnAsserter {
    public OrderIdColumnAsserter(IGridPage gridPage, WebDriver driver) {
        super(gridPage, driver);
    }

    @Test
    public void orderIdEqualToFilter() throws Exception {
        getGridPage().navigateTo();
        var newItem = createNewItemInDb();

        getGridPage().getGrid().filter(GridColumns.ORDER_ID, FilterOperator.EQUAL_TO, newItem.getOrderId());
        waitForGridToLoad(1, getGridPage().getGrid());
        List<GridItem> items = getGridPage().getGrid().getItems();

        Assert.assertEquals(items.stream().count(), 1);
    }

    public void orderIdGreaterThanOrEqualToFilter() throws Exception {
        getGridPage().navigateTo();
        // Create new item with unique ship name;
        var newItem = createNewItemInDb();

        // Create second new item with the same unique shipping name
        var secondNewItem = createNewItemInDb(newItem.getShipName());

        // When we filter by the second unique column ShippingName, only one item will be displayed. Once we apply the second not equal to filter the grid should be empty.
        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.GREATER_THAN_OR_EQUAL_TO, newItem.getOrderId()),
                new GridFilter(GridColumns.SHIP_NAME, FilterOperator.EQUAL_TO, newItem.getShipName()));

        waitForGridToLoadAtLeast(2, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertEquals(results.stream().filter(x -> x.getShipName() == secondNewItem.getShipName()).findFirst().get().getOrderId(), secondNewItem.getOrderId());
        Assert.assertEquals(results.stream().filter(x -> x.getShipName() == newItem.getShipName()).findFirst().get().getOrderId(), newItem.getOrderId());
        Assert.assertTrue(results.stream().count() == 2);
    }

    public void orderIdGreaterThanFilter() throws Exception {
        getGridPage().navigateTo();
        // Create new item with unique ship name;
        var newItem = createNewItemInDb();

        // Create second new item with the same unique shipping name
        var secondNewItem = createNewItemInDb(newItem.getShipName());

        // Filter by the smaller orderId but also by the second unique column in this case shipping name
        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.GREATER_THAN, newItem.getOrderId()),
                new GridFilter(GridColumns.SHIP_NAME, FilterOperator.EQUAL_TO, newItem.getShipName()));
        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertEquals(results.stream().filter(x -> x.getShipName() == secondNewItem.getShipName()).findFirst().get().getOrderId(), secondNewItem.getOrderId());
        Assert.assertTrue(results.stream().count() == 1);
    }

    public void orderIdLessThanOrEqualToFilter() throws Exception {
        getGridPage().navigateTo();
        // Create new item with unique ship name;
        var newItem = createNewItemInDb();

        // Create second new item with the same unique shipping name
        var secondNewItem = createNewItemInDb(newItem.getShipName());

        // Filter by the larger orderId but also by the second unique column in this case shipping name
        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.LESS_THAN_OR_EQUAL_TO, secondNewItem.getOrderId()),
                new GridFilter(GridColumns.SHIP_NAME, FilterOperator.EQUAL_TO, newItem.getShipName()));
        waitForGridToLoadAtLeast(2, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertEquals(results.stream().filter(x -> x.getShipName() == newItem.getShipName()).findFirst().get().getOrderId(), newItem.getOrderId());
        Assert.assertEquals(results.stream().filter(x -> x.getShipName() == secondNewItem.getShipName()).skip(results.stream().count() - 1).findFirst().get().getOrderId(), secondNewItem.getOrderId());
        Assert.assertTrue(results.stream().count() == 2);
    }

    public void orderIdLessThanFilter() throws Exception {
        getGridPage().navigateTo();
        // Create new item with unique ship name;
        var newItem = createNewItemInDb();

        // Create second new item with the same unique shipping name
        var secondNewItem = createNewItemInDb(newItem.getShipName());

        // Filter by the larger orderId but also by the second unique column in this case shipping name
        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.LESS_THAN, secondNewItem.getOrderId()),
                new GridFilter(GridColumns.SHIP_NAME, FilterOperator.EQUAL_TO, secondNewItem.getShipName()));
        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertEquals(results.stream().filter(x -> x.getShipName() == newItem.getShipName()).findFirst().get().getOrderId(), newItem.getOrderId());
        Assert.assertTrue(results.stream().count() == 1);
    }

    public void orderIdNotEqualToFilter() throws Exception {
        getGridPage().navigateTo();
        // Create new item with unique ship name;
        var newItem = createNewItemInDb();

        // Create second new item with the same unique shipping name
        var secondNewItem = createNewItemInDb(newItem.getShipName());

        // Filter by the larger orderId but also by the second unique column in this case shipping name
        getGridPage().getGrid().filter(
                new GridFilter(GridColumns.ORDER_ID, FilterOperator.NOT_EQUAL_TO, secondNewItem.getOrderId()),
                new GridFilter(GridColumns.SHIP_NAME, FilterOperator.EQUAL_TO, secondNewItem.getShipName()));
        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertEquals(results.stream().filter(x -> x.getShipName() == newItem.getShipName()).findFirst().get().getOrderId(), newItem.getOrderId());
        Assert.assertTrue(results.stream().count() == 1);
    }

    public void orderIdClearFilter() throws Exception {
        getGridPage().navigateTo();
        // Create new item with unique ship name;
        var newItem = createNewItemInDb();

        // Make sure that we have at least 2 items if the grid is empty. The tests are designed to run against empty DB.
        var secondNewItem = createNewItemInDb(newItem.getShipName());

        getGridPage().getGrid().filter(GridColumns.ORDER_ID, FilterOperator.EQUAL_TO, newItem.getOrderId());
        waitForGridToLoad(1, getGridPage().getGrid());
        getGridPage().getGrid().removeFilters();

        waitForGridToLoadAtLeast(1, getGridPage().getGrid());
        List<Order> results = getGridPage().getGrid().getItems();

        Assert.assertTrue(results.stream().count() > 1);
    }
}
